website/blog/2021-02-04-raptorx.md

```diff
@@ -0,0 +1,152 @@
+
+
+
+The following figure shows the architecture of the caching solution. We have hierarchical layers of cache, which will be introduced in details in the remainder of this note:
+
+* Metastore versioned cache: We cache table/partition information in coordinator. Given metadata is mutable, like Iceberg or Delta Lake, the information is versioned. We only sync versions with metastore and fetch updated metadata when version is out of date.
+* File list cache: Cache file list from remote storage partition directory.
+* Fragment result cache: Cache partially computed result on leaf workers’ local SSDs. Pruning techniques are required to simplify query plans as they change all the time.
+* File handle and footer cache: Cache open file descriptor and stripe/file footer information in leaf worker memory. These pieces of data are mostly frequently accessed when reading files.
+* Alluxio data cache: Cache file segments with 1MB aligned chunks on leaf workers’ local SSDs. The library is built from Alluxio cache service.
+* Affinity scheduler: A scheduler that sends sticky requests based on file paths to particular workers to maximize cache hit rate.
+
+
+
+![RaptorX Architecture](/img/blog/2021-02-04-raptorx/raptorx-arch.png)
+
+
+## Metastore Versioned Cache
+
+A Presto coordinator caches table metadata (schema, partition list, and partition info) to avoid long `getPartitions` calls to Hive Metastore. However, Hive table metadata is mutable. Versioning is needed to determine if the cached metadata is valid or not. To achieve that, the coordinator attaches a version number to each cache key-value pair. When a read request comes, the coordinator asks the Hive Metastore to get the partition info (if it's not cached at all) or confirm with Hive Metastore to check if cached info is up to date. Though the roundtrip to Hive Metastore cannot be avoided, the version matching is relatively cheap compared with fetching the entire partition info.
+
+
+
+## File List cache
+
+A Presto coordinator caches file lists in memory to avoid long `listFile` calls to remote storage. This can only be applied to sealed directories. For open partitions, Presto will skip caching those directories to guarantee data freshness. One major use case for open partitions is to support the need of near-realtime ingestion and serving. In such cases, the ingestion engines (e.g., micro batch) will keep writing new files to the open partitions so that Presto can read near-realtime data. Further details like compaction, metastore update, or replication for near-realtime ingestion will be out of the scope of this note.
+
+
+
+## Fragment Result Cache
+
+A Presto worker that is running a leaf stage can decide if to cache the partially computed results on local SSD. This is to prevent duplicated computation upon multiple queries. The most typical use case is to cache the plan fragments on leaf stage with one level of scan, filter, project, or/and aggregation.
```

Comment:
```suggestion
A Presto worker that is running a leaf stage can decide to cache the partially computed results on local SSD. This is to prevent duplicated computation upon multiple queries. The most typical use case is to cache the plan fragments on leaf stage with one level of scan, filter, project, and/or aggregation.
```

Swapping and/or is me being weird, but when they are together like that it's usually in that order. There isn't an actual rule for that :-P